src/Common/Fiber.h

```diff
@@ -4,4 +4,181 @@
+    Fiber(StackAlloc && salloc, Fn && fn) : impl(std::allocator_arg_t(), std::forward<StackAlloc>(salloc), std::forward<Fn>(fn))
+    {
+    }
+
+    Fiber() = default;
+
+    Fiber(Fiber && other) = default;
+    Fiber & operator=(Fiber && other) = default;
+
+    Fiber(const Fiber &) = delete;
+    Fiber & operator =(const Fiber &) = delete;
+
+    ~Fiber()
+    {
+        for (auto & [_, data] : local_data)
+            (*data.cleanup_fn)(data.ptr);
+    }
+
+    explicit operator bool() const
+    {
+        return impl.operator bool();
+    }
+
+    void resume()
+    {
+        /// Update information about current executing fiber.
+        auto & current_fiber_info = getCurrentFiberInfo();
+        auto parent_fiber_info = current_fiber_info;
+        current_fiber_info = FiberInfo{this, &parent_fiber_info};
+        impl = std::move(impl).resume();
+        /// Restore current fiber info.
+        current_fiber_info = parent_fiber_info;
+    }
+
+    /// Set pointer to fiber-specific data, it will be stored in hash-map
+    /// using provided key and cleaned on fiber destructor using provided
+    /// cleanup function.
+    void setLocalData(void * key, void * ptr, CleanUpFn * cleanup_fn)
+    {
+        local_data[key] = FiberLocalData{ptr, cleanup_fn};
+    }
+
+    /// Get pointer to fiber-specific data by key.
+    /// If no data was stored by this key, return nullptr.
+    void * getLocalData(void * key)
+    {
+        return local_data[key].ptr;
+    }
+
+    struct FiberInfo
+    {
+        Fiber * fiber = nullptr;
+        FiberInfo * parent_info = nullptr;
+    };
+
+    static FiberInfo & getCurrentFiberInfo()
+    {
+        thread_local static FiberInfo current_fiber_info;
+        return current_fiber_info;
+    }
+
+private:
+    struct FiberLocalData
+    {
+        void * ptr;
+        CleanUpFn * cleanup_fn;
+    };
+
+    Impl impl;
+    std::unordered_map<void *, FiberLocalData> local_data;
+};
+
+/// Implementation for fiber local variable.
+/// If we are not in fiber, it returns thread local data.
+/// If we are in fiber, it returns fiber local data.
+/// Fiber local data is destroyed in Fiber destructor.
+/// On first request, fiber local data is copied from parent
+/// fiber data or from current thread data if there is no parent fiber.
+/// Implementation is similar to boost::fiber::fiber_specific_ptr
+/// (we cannot use it because we don't use boost::fiber API.
+template <typename T>
+class FiberLocal
+{
+public:
+    struct DefaultCleanUpFn : public Fiber::CleanUpFn
+    {
+        void operator()(void * data) override
+        {
+            delete static_cast<T *>(data);
+        }
+    };
+
+    struct CustomCleanUpFn : public Fiber::CleanUpFn
+    {
+        explicit CustomCleanUpFn(void (*fn_)(T*)) : fn(fn_)
+        {
+        }
+
+        void operator()(void * data) override
+        {
+            if (likely(fn != nullptr))
+                fn(static_cast<T *>(data));
```

Comment:
Maybe if we want this, we'd better call `delete static_cast<T *>(data);` here (not expect fn will do it)